File: backend/models/analysis.py
from pydantic import BaseModel
from typing import Literal, Optional
from datetime import datetime
from enum import Enum


class Language(str, Enum):
    PYTHON = "python"
    JAVASCRIPT = "javascript"


class Severity(str, Enum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"
    INFO = "info"


class IssueStatus(str, Enum):
    VERIFIED = "verified"
    POTENTIAL = "potential"
    SUGGESTION = "suggestion"


class AnalysisOptions(BaseModel):
    max_iterations: int = 3
    include_style_checks: bool = True
    generate_fixes: bool = True


class TestResult(BaseModel):
    test_code: str
    passed: bool
    stdout: str
    stderr: str
    exit_code: int
    duration_ms: int
    error_line: Optional[int] = None


class Issue(BaseModel):
    id: str
    title: str
    description: str
    severity: Severity
    status: IssueStatus
    line_start: int
    line_end: int
    test_result: Optional[TestResult] = None
    suggested_fix: Optional[str] = None
    reasoning: str


class AgentStep(BaseModel):
    step_number: int
    action: Literal["analyzing", "generating_test", "executing", "refining", "complete"]
    description: str
    timestamp: datetime
    data: Optional[dict] = None


class AnalysisResult(BaseModel):
    issues: list[Issue]
    summary: str
    code_quality_score: int
    steps: list[AgentStep]
    total_duration_ms: int


class AnalysisSession(BaseModel):
    id: str
    code: str
    language: Language
    status: Literal["pending", "running", "completed", "error"]
    result: Optional[AnalysisResult] = None
    created_at: datetime
    completed_at: Optional[datetime] = None


class AnalyzeRequest(BaseModel):
    code: str
    language: Language
    options: Optional[AnalysisOptions] = None


class AnalyzeResponse(BaseModel):
    session_id: str
    status: Literal["started"]

File: frontend/src/types/analysis.ts
export type Language = 'python' | 'javascript'
export type Severity = 'critical' | 'high' | 'medium' | 'low' | 'info'
export type IssueStatus = 'verified' | 'potential' | 'suggestion'
export type SessionStatus = 'pending' | 'running' | 'completed' | 'error'
export type StepAction = 'analyzing' | 'generating_test' | 'executing' | 'refining' | 'complete'

export interface TestResult {
  test_code: string
  passed: boolean
  stdout: string
  stderr: string
  exit_code: number
  duration_ms: number
  error_line?: number
}

export interface Issue {
  id: string
  title: string
  description: string
  severity: Severity
  status: IssueStatus
  line_start: number
  line_end: number
  test_result?: TestResult
  suggested_fix?: string
  reasoning: string
}

export interface AgentStep {
  step_number: number
  action: StepAction
  description: string
  timestamp: string
  data?: Record<string, unknown>
}

export interface AnalysisResult {
  issues: Issue[]
  summary: string
  code_quality_score: number
  steps: AgentStep[]
  total_duration_ms: number
}

export interface AnalysisSession {
  id: string
  code: string
  language: Language
  status: SessionStatus
  result?: AnalysisResult
  created_at: string
  completed_at?: string
}

export type Strictness = 'relaxed' | 'normal' | 'strict'

export interface AnalyzeRequest {
  code: string
  language: Language
  strictness?: Strictness
  max_issues?: number
}

export interface AnalyzeResponse {
  session_id: string
  status: 'started'
}
